README.md
Android Native Development Kit (NDK)
====================================

The latest version of this document is available at
https://android.googlesource.com/platform/ndk/+/master/README.md.

**Note:** This document is for developers _of_ the NDK, not developers that use
the NDK.

The NDK allows Android application developers to include native code in their
Android application packages, compiled as JNI shared libraries.

[TOC]

Other Resources
---------------

 * User documentation is available on the [Android Developer website].
 * Discussions related to the Android NDK happen on the [android-ndk Google
   Group].
 * Announcements such as releases are posted to the [android-ndk-announce Google
   Group].
 * File bugs against the NDK at https://github.com/android-ndk/ndk/issues.

[Android Developer website]: https://developer.android.com/ndk/index.html
[android-ndk Google Group]: http://groups.google.com/group/android-ndk
[android-ndk-announce Google Group]: http://groups.google.com/group/android-ndk-announce

Building the NDK
================

Both Linux and Windows NDKs are built on Linux machines. Windows host binaries
are cross-compiled with MinGW.

Building the NDK for Mac OS X requires at least 10.8.

Components
----------

The NDK components can be loosely grouped into host toolchains, target
prebuilts, build systems, and support libraries.

### Host Toolchains

* `toolchains/` contains GCC and Clang toolchains.
    * `$TOOLCHAIN/config.mk` contains ARCH and ABIS this toolchain can handle.
    * `$TOOLCHAIN/setup.mk` contains toolchain-specific default CFLAGS/LDFLAGS
      when this toolchain is used.
* `prebuilt/$HOST_TAG` contains build dependencies and additional tools.
    * make, python, yasm, and for Windows: cmp.exe and echo.exe
    * `ndk-depends`, `ndk-stack` and `ndk-gdb` can also be found here.

### Target Prebuilts

* `sysroot/usr/include` contains the headers for the NDK. See [Unified Headers]
  for more information.
* `platforms/android-$VERSION/arch-$ARCH_NAME/` contains stub shared libraries
  and a few static libraries for each API level. See [Platform APIs] for more
  information.
* `sources/cxx-stl/$STL` contains the headers and libraries for the various C++
  STLs.
* `prebuilt/android-$ARCH/gdbserver` contains gdbserver.

[Unified Headers]: docs/UnifiedHeaders.md
[Platform APIs]: docs/PlatformApis.md

### Build Systems

* `build/` contains ndk-build, the NDK's home grown build system. Most of the
  implementation lives in `build/core`.
* `build/cmake` contains components for using the NDK with CMake (at present
  only a CMake toolchain file, but in the future it will contain CMake modules
  that CMake will load, obviating the need for a toolchain file).
* `build/tools` contains `make_standalone_toolchain.py`, but also contains
  legacy sripts that were used to build the NDK. Eventually, this should contain
  nothing but the standalone toolchain scripts.
* The gradle plugins for the NDK are not included in the NDK.

### Support Libraries

* `sources/android` and `sources/third_party` contain modules that can be used
  in apps (gtest, cpufeatures, native\_app\_glue, etc) via
  `$(call import-module,$MODULE)`.

Prerequisites
-------------

* [AOSP NDK Repository](http://source.android.com/source/downloading.html)
    * Check out the branch `master-ndk`

        ```bash
        repo init -u https://android.googlesource.com/platform/manifest \
            -b master-ndk

        # Googlers, use
        repo init -u \
            persistent-https://android.git.corp.google.com/platform/manifest \
            -b master-ndk
        ```

Linux dependencies are listed in the [Dockerfile]. You can use docker to build
the NDK:

```bash
docker build -t ndk-dev infra/docker
docker run -it -u $UID -v `realpath ..`:/src -w /src/ndk ndk-dev ./checkbuild.py
```

Building on Mac OS X has similar dependencies as Linux, but also requires Xcode.

Running tests requires that `adb` is in your `PATH`. This is provided as part of
the [Android SDK].

[Dockerfile]: infra/docker/Dockerfile
[Android SDK]: https://developer.android.com/studio/index.html#downloads

Building the NDK
----------------

### For Linux or Darwin:

```bash
$ python checkbuild.py
```

### For Windows, from Linux:

```bash
$ python checkbuild.py --system windows  # Or windows64.
```

`checkbuild.py` will also build all of the NDK tests. This takes about four
times as long as building the NDK itself, so pass `--no-build-tests` to skip
building the tests. They can be built later with `python run_tests.py
--rebuild`.

`checkbuild.py` also accepts a variety of other options to speed up local
builds, namely `--arch` and `--module`.

Packaging
---------

By default, `checkbuild.py` will also package the NDK. To skip the packaging
step, use the `--no-package` flag. To avoid packaging an incomplete NDK,
packaging will not be run if `--module` was passed unless `--force-package` was
also provided.


CHANGELOG.md
Changelog
=========

Report issues to [GitHub].

For Android Studio issues, follow the docs on the [Android Studio site].

[GitHub]: https://github.com/android-ndk/ndk/issues
[Android Studio site]: http://tools.android.com/filing-bugs

Announcements
-------------

 * GCC is no longer supported. It will be removed in NDK r18.

 * `libc++` is now the default STL for CMake and standalone toolchains. If you
   manually selected a different STL, we strongly encourage you to move to
   `libc++`. Note that ndk-build still defaults to no STL. For more details, see
   [this blog post](https://android-developers.googleblog.com/2017/09/introducing-android-native-development.html).

 * gnustl and stlport are deprecated and will be removed in NDK r18.

 * Support for ARMv5 (armeabi) is deprecated. It will no longer build by default
   with ndk-build, but is still buildable if it is explicitly named, and will be
   included by "all" and "all32". Support for ARMv5 has been removed in r17.

   Both CMake and ndk-build will issue a warning if you target this ABI.

 * Support for ICS (android-14 and android-15) will be removed from r18.

 * The Play Store will require 64-bit support when uploading an APK beginning in
   August 2019. Start porting now to avoid surprises when the time comes. For
   more information, see [this blog post](https://android-developers.googleblog.com/2017/12/improving-app-security-and-performance.html).

Changes
-------

 * Updated Clang to build 4691093, based on r316199.
     * [Issue 313]: LTO now works on Windows.
     * [Issue 527]: [Undefined Behavior Sanitizer] now works for NDK binaries.
     * [Issue 593]: Clang now automatically passes `--eh-frame-hdr` to the
       linker even for static executables, fixing exception handling.
 * Updated gtest to upstream revision 0fe96607d85cf3a25ac40da369db62bbee2939a5.
 * `libandroid_support` is no longer used when your NDK API level is greater
   than or equal to 21 (Lollipop). Build system maintainers: be sure to update
   your build systems to account for this.
 * The platform static libraries (libc.a, libm.a, etc.) have been updated.
     * All NDK platforms now contain a modern version of these static libraries.
       Previously they were all Gingerbread (perhaps even older) or Lollipop.
     * Prior NDKs could not use the static libraries with a modern NDK API level
       because of symbol collisions between libc.a and libandroid_support. This
       has been solved by removing libandroid_support for modern API levels. A
       side effect of this is that you must now target at least android-21 to
       use the static libraries, but these binaries will still work on older
       devices.
 * Fixed parsing of the NDK revision in CMake. NDK version information is now
   available in the following CMake variables:
     * `ANDROID_NDK_REVISION`: The full string in the source.properties file.
     * `ANDROID_NDK_MAJOR`: The major revision of the NDK. For example: the 16
       in r16b.
     * `ANDROID_NDK_MINOR`: The minor revision of the NDK. For example: the b
       (represented as 1) in r16b.
     * `ANDROID_NDK_BUILD`: The build number of the NDK. This is 0 in the case
       of a local development build.
     * `ANDROID_NDK_BETA`: The beta version of the NDK. This is 0 for a stable
       release.
 * Added support for `APP_WRAP_SH` to ndk-build.
     * This variable points to a shell script (relative to your Android.mk) that
       will be installed as a [wrap.sh] file in your APK.
     * Available in both an ABI-generic form (`APP_WRAP_SH`), which will install
       a single script for every ABI, and an ABI-specific form
       (`APP_WRAP_SH_arm64-v8a`, etc) to allow for per-ABI customization of the
       wrap.sh script.
 * [Issue 540]: ndk-build now installs sanitizer runtime libraries to your out
   directory for inclusion in your APK. Coupled with [wrap.sh], this removes the
   requirement of rooting your device to use sanitizers.
 * When using ASAN, ndk-build will install a wrap.sh file to set up ASAN for
   your app if you have not specified your own wrap.sh. If you have specified
   your own wrap.sh, you can add ASAN support to it as described
   [here](https://github.com/google/sanitizers/wiki/AddressSanitizerOnAndroidO).
 * [Issue 614]: ndk-build and CMake no longer pass `-fno-integrated-as` for
   arm32 by default.  If your project contains hand written assembly, you may
   need to pass this flag.
 * ion.h is no longer a part of the NDK. This file defined an API that is not
   consistent across devices.

[Undefined Behavior Sanitizer]: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
[wrap.sh]: https://developer.android.com/ndk/guides/wrap-script.html
[Issue 313]: https://github.com/android-ndk/ndk/issues/313
[Issue 527]: https://github.com/android-ndk/ndk/issues/527
[Issue 540]: https://github.com/android-ndk/ndk/issues/540
[Issue 593]: https://github.com/android-ndk/ndk/issues/593
[Issue 614]: https://github.com/android-ndk/ndk/issues/614

Known Issues
------------

 * This is not intended to be a comprehensive list of all outstanding bugs.
 * [Issue 360]: `thread_local` variables with non-trivial destructors will cause
   segfaults if the containing library is `dlclose`ed on devices running M or
   newer, or devices before M when using a static STL. The simple workaround is
   to not call `dlclose`.
 * [Issue 70838247]: Gold emits broken debug information for AArch64. AArch64
   still uses BFD by default.

[Issue 360]: https://github.com/android-ndk/ndk/issues/360
[Issue 70838247]: https://issuetracker.google.com/70838247

* [Issue 399]: MIPS64 must use the integrated assembler. Clang defaults to
   using binutils rather than the integrated assembler for this target.
   ndk-build and cmake handle this for you, but make sure to use
   `-fintegrated-as` for MIPS64 for custom build systems.

[Issue 399]: https://github.com/android-ndk/ndk/issues/399


validate.py
#!/usr/bin/env python
#
# Copyright (C) 2016 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Runs the test suite over a set of devices."""
from __future__ import print_function

import argparse
import distutils.spawn
import logging
import os
import re
import shutil
import signal
import site
import subprocess
import sys
import yaml

import ndk.debug
import ndk.notify
import ndk.paths


THIS_DIR = os.path.realpath(os.path.dirname(__file__))


class Device(object):
    def __init__(self, serial, name, version, build_id, abis, is_emulator):
        self.serial = serial
        self.name = name
        self.version = version
        self.build_id = build_id
        self.abis = abis
        self.is_emulator = is_emulator

    def __str__(self):
        return 'android-{} {} {} {}'.format(
            self.version, self.name, self.serial, self.build_id)


class DeviceFleet(object):
    def __init__(self, test_configurations):
        """Initializes a device fleet.

        Args:
            test_configurations: Dict mapping API levels to a list of ABIs to
                test for that API level. Example:

                    {
                        15: ['armeabi', 'armeabi-v7a'],
                        16: ['armeabi', 'armeabi-v7a', 'x86'],
                    }
        """
        self.devices = {}
        for api, abis in test_configurations.items():
            self.devices[api] = {abi: None for abi in abis}

    def add_device(self, device):
        if device.version not in self.devices:
            print('Ignoring device for unwanted API level: {}'.format(device))
            return

        same_version = self.devices[device.version]
        for abi, current_device in same_version.iteritems():
            # This device can't fulfill this ABI.
            if abi not in device.abis:
                continue

            # Never houdini.
            if abi.startswith('armeabi') and 'x86' in device.abis:
                continue

            # Anything is better than nothing.
            if current_device is None:
                self.devices[device.version][abi] = device
                continue

            # The emulator images have actually been changed over time, so the
            # devices are more trustworthy.
            if current_device.is_emulator and not device.is_emulator:
                self.devices[device.version][abi] = device

    def get_device(self, version, abi):
        return self.devices[version][abi]

    def get_missing(self):
        missing = []
        for version, abis in self.devices.iteritems():
            for abi, device in abis.iteritems():
                if device is None:
                    missing.append('android-{} {}'.format(version, abi))
        return missing

    def get_versions(self):
        return self.devices.keys()

    def get_abis(self, version):
        return self.devices[version].keys()


def get_device_abis(properties):
    # 64-bit devices list their ABIs differently than 32-bit devices. Check all
    # the possible places for stashing ABI info and merge them.
    abi_properties = [
        'ro.product.cpu.abi',
        'ro.product.cpu.abi2',
        'ro.product.cpu.abilist',
    ]
    abis = set()
    for abi_prop in abi_properties:
        if abi_prop in properties:
            abis.update(properties[abi_prop].split(','))

    return sorted(list(abis))


def get_device_details(serial):
    import adb  # pylint: disable=import-error
    props = adb.get_device(serial).get_props()
    name = props['ro.product.name']
    version = int(props['ro.build.version.sdk'])
    supported_abis = get_device_abis(props)
    build_id = props['ro.build.id']
    is_emulator = props.get('ro.build.characteristics') == 'emulator'
    return Device(serial, name, version, build_id, supported_abis, is_emulator)


def find_devices(sought_devices):
    """Detects connected devices and returns a set for testing.

    We get a list of devices by scanning the output of `adb devices`. We want
    to run the tests for the cross product of the following configurations:

    ABIs: {armeabi, armeabi-v7a, arm64-v8a, mips, mips64, x86, x86_64}
    Platform versions: {android-10, android-16, android-21}
    Toolchains: {clang, gcc}

    Note that not all ABIs are available for every platform version. There are
    no 64-bit ABIs before android-21, and there were no MIPS ABIs for
    android-10.
    """
    if distutils.spawn.find_executable('adb') is None:
        raise RuntimeError('Could not find adb.')

    # We could get the device name from `adb devices -l`, but we need to
    # getprop to find other details anyway, and older devices don't report
    # their names properly (nakasi on android-16, for example).
    p = subprocess.Popen(['adb', 'devices'], stdout=subprocess.PIPE)
    out, _ = p.communicate()
    if p.returncode != 0:
        raise RuntimeError('Failed to get list of devices from adb.')

    # The first line of `adb devices` just says "List of attached devices", so
    # skip that.
    fleet = DeviceFleet(sought_devices)
    for line in out.split('\n')[1:]:
        if not line.strip():
            continue

        serial, _ = re.split(r'\s+', line, maxsplit=1)

        if 'offline' in line:
            print('Ignoring offline device: ' + serial)
            continue
        if 'unauthorized' in line:
            print('Ignoring unauthorized device: ' + serial)
            continue

        device = get_device_details(serial)
        print('Found device {}'.format(device))
        fleet.add_device(device)

    return fleet


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'ndk', metavar='NDK', type=os.path.realpath, nargs='?',
        help='NDK to validate. Defaults to ../out/android-ndk-$RELEASE.')

    parser.add_argument(
        '--abi', dest='abis', action='append',
        help=('ABIs to test. Defaults to all available for the connected '
              'devices.'))
    parser.add_argument(
        '--platform', dest='platforms', action='append', type=int,
        help=('Device API levels to test. Must be a subset of '
              'DeviceFleet.devices.'))
    parser.add_argument(
        '--toolchain', dest='toolchains', action='append',
        choices=('clang', '4.9'), help='Toolchains to test.')
    parser.add_argument(
        '--headers-config', dest='headers_configs', action='append',
        choices=('unified', 'deprecated'), help='Headers configs to test.')

    import tests.testlib
    parser.add_argument(
        '--suite', dest='suites', action='append',
        choices=tests.testlib.ALL_SUITES, help='Test suites to run.')

    parser.add_argument(
        '--filter', help='Only run tests that match the given pattern.')
    parser.add_argument(
        '--log-dir', type=os.path.realpath, default='test-logs',
        help='Directory to store test logs.')
    parser.add_argument(
        '--config', type=os.path.realpath, default='qa_config.yaml',
        help='Path to the config file describing the test run.')
    parser.add_argument(
        '-v', '--verbose', action='count', default=0,
        help='Increase log level. Defaults to logging.WARNING.')

    return parser.parse_args()


def get_aggregate_results(details):
    tests = {}
    for config, report in details.iteritems():
        for suite, suite_report in report.by_suite().items():
            for test_report in suite_report.all_failed:
                name = '.'.join([suite, test_report.result.test.name])
                if name not in tests:
                    tests[name] = []
                tests[name].append((config, test_report.result))
    return tests


def print_aggregate_details(details, use_color):
    tests = get_aggregate_results(details)
    for test_name, configs in tests.iteritems():
        # We might be printing a lot of crap here, so let's be obvious about
        # where each test starts.
        print('BEGIN TEST RESULT: ' + test_name)
        print('=' * 80)

        for config, result in configs:
            print('FAILED {}:'.format(config))
            print(result.to_string(colored=use_color))


def main():
    if sys.platform != 'win32':
        ndk.debug.register_debug_handler(signal.SIGUSR1)
        ndk.debug.register_trace_handler(signal.SIGUSR2)

    args = parse_args()

    log_levels = [logging.WARNING, logging.INFO, logging.DEBUG]
    verbosity = min(args.verbose, len(log_levels) - 1)
    log_level = log_levels[verbosity]
    logging.basicConfig(level=log_level)

    os.chdir(THIS_DIR)

    if args.ndk is None:
        args.ndk = ndk.paths.get_install_path()

    # We need to do this here rather than at the top because we load the module
    # from a path that is given on the command line. We load it from the NDK
    # given on the command line so this script can be run even without a full
    # platform checkout.
    site.addsitedir(os.path.join(args.ndk, 'python-packages'))

    ndk_build_path = os.path.join(args.ndk, 'ndk-build')
    if os.name == 'nt':
        ndk_build_path += '.cmd'
    if not os.path.exists(ndk_build_path):
        sys.exit(ndk_build_path + ' does not exist.')

    with open(args.config) as test_config_file:
        test_config = yaml.load(test_config_file)

    fleet = find_devices(test_config['devices'])
    print('Test configuration:')
    for version in sorted(fleet.get_versions()):
        print('\tandroid-{}:'.format(version))
        for abi in sorted(fleet.get_abis(version)):
            print('\t\t{}: {}'.format(abi, fleet.get_device(version, abi)))
    missing_configs = fleet.get_missing()
    if len(missing_configs):
        print('Missing configurations: {}'.format(', '.join(missing_configs)))

    if os.path.exists(args.log_dir):
        shutil.rmtree(args.log_dir)
    os.makedirs(args.log_dir)

    if args.headers_configs is None:
        headers_configs = args.headers_configs
    else:
        headers_configs = [c == 'deprecated' for c in args.headers_configs]

    use_color = sys.stdin.isatty() and os.name != 'nt'
    with ndk.paths.temp_dir_in_out('validate-out') as out_dir:
        import tests.runners
        good, details = tests.runners.run_for_fleet(
            args.ndk, fleet, out_dir, args.filter, args.platforms, args.abis,
            args.toolchains, headers_configs, args.suites, use_color)

    print_aggregate_details(details, use_color)

    subject = 'NDK Testing {}!'.format('Passed' if good else 'Failed')
    ndk.notify.toast(subject)

    sys.exit(not good)


if __name__ == '__main__':
    main()
